blackjax/inference/nuts.py

```diff
@@ -0,0 +1,115 @@
+    def do_keep_expanding(trajectory, proposal, depth):
+        # check if tree is turning
+        # divergence should only be checked here or below.
+        return depth < max_tree_depth
+
+    def expand(trajectory, proposal, depth):
+
+        # create new subtrajectory twice as long as previous
+        direction = jax.random.bernoulli(rng_key)
+        state = jax.lax.cond(
+            direction > 0,
+            trajectory,
+            lambda trajectory: trajectory.rightmost_state,
+            trajectory,
+            lambda trajectory: trajectory.leftmost_state,
+        )
+        state, subtrajectory, new_proposal = create_subtrajectory(state, trajectory, proposal, 2 * depth)
+
+        # merge the subtrajectory to the trajectory
+        leftmost_state, rightmost_state = jax.lax.cond(
+            direction > 0,
+            trajectory,
+            lambda trajectory: (trajectory.leftmost_state, state),
+            trajectory,
+            lambda trajectory: (state, trajectory.rightmost_state),
+        )
+        new_trajectory = Trajectory(leftmost_state, rightmost_state)
+
+        # update the proposal
+        p_update = progressive_biased_sampling(trajectory, subtrajectory)
+        do_update_proposal = jax.random.bernoulli(rng_key, p_update)
+        new_proposal = jax.lax.cond(
+            do_update_proposal,
+            new_proposal,
+            lambda x: x,
+            proposal,
+            lambda x: x,
+        )
+
+        return new_trajectory, new_proposal, depth + 1
+
+    trajectory, proposal, _ = jax.while_loop(do_keep_expanding, expand, state)
+    return proposal
+
+
+def create_subtrajectory(state, direction, max_length):
+    num_steps = 0
+
+    def append():
+        pass
+
+    def do_keep_expanding(state, trajectory, proposal):
+        # update checkpoints
+        # check turning condition
+        return num_steps < max_length
+
+    def expand(state, trajectory, proposal):
+        """We concurrently expand the trajectory and draw a new
+        proposal from this trajectory.
+        """
+        new_state = integrator(state, direction * step_size)
+        new_trajectory = append(new_state, trajectory, direction)
+
+        p_update = progressive_uniform_sampling(trajectory, new_state)
+        do_update_proposal = jax.random.bernoulli(rng_key, p_update)
+        new_proposal = jax.lax.cond(
+            do_update_proposal,
+            new_state,
+            lambda x: x,
+            proposal,
+            lambda x: x,
+        )
+
+        return new_state, new_trajectory, new_proposal
+
+    trajectory = jax.while_cond(do_keep_expanding, expand, state)
+    return trajectory
+
+
+def progressive_uniform_sampling(trajectory, new_state):
+    p_accept = jax.scipy.special.expit(trajectory.weight, new_state.energy)
+    return p_accept
+
+
+def progressive_biased_sampling(trajectory, new_trajectory):
+    """numpyro includes the turning and diverging in this check but I don't
```

Comment:
whatever where you put turning/diverging check, just making sure that you won't propose a point in a turning/diverging new_trajectory. This is necessary for detailed balance.